docs: Add ProcessWire to the Quickstart List

BernhardBaumrock opened the pull request:
## The Issue

Add Processwire CMS to quickstarts

## How This PR Solves The Issue

* Add the quickstart
* Add the automated test

## Manual Testing Instructions

Review rendered docs at https://ddev--6879.org.readthedocs.build/en/6879/users/quickstart/#processwire

* Step through both types of quickstart and make sure they work right
* Review the automated test and make sure it properly tests the quickstart

## Automated Testing Overview

Added processwire.bats.

tyler36 commented on docs/content/users/quickstart.md:
````diff
@@ -628,6 +628,30 @@ The Laravel project type can be used for [StarterKits](https://laravel.com/docs/
+
+    ```bash
+    mkdir my-processwire-site && cd my-processwire-site
+    curl -LJO https://github.com/processwire/processwire/archive/master.zip
+    unzip ./processwire-master.zip && rm -f processwire-master.zip && mv -f ./processwire-master/{.,}* . ; rm -rf processwire-master
+    ddev config --webserver-type=apache-fpm
````
```suggestion
    ddev config --webserver-type=apache-fpm
    ddev composer create --prefer-dist --no-interaction processwire/processwire
```

Most of the currently supported frameworks install via composer, where available. This helps with php and vendor constrainsts.

tyler36 reviewed (commented):
Thank you for the framework suggestion.

To keep consistent, please consider using the `composer` installations suggested.

Note: I'm not familiar with ProcessWire but the suggestions result in the same configuration pages displayed in the original PR.

BernhardBaumrock commented on docs/content/users/quickstart.md:
````diff
@@ -628,6 +628,30 @@ The Laravel project type can be used for [StarterKits](https://laravel.com/docs/
+
+    ```bash
+    mkdir my-processwire-site && cd my-processwire-site
+    curl -LJO https://github.com/processwire/processwire/archive/dev.zip
+    unzip ./processwire-dev.zip && rm -f processwire-dev.zip && mv -f ./processwire-dev/{.,}* . ; rm -rf processwire-dev
+    ddev config --webserver-type=apache-fpm
````
@tyler36 thx but that takes a lot longer than my code and I get the following warnings:

<img width="928" alt="image" src="https://github.com/user-attachments/assets/b2a3a298-6dec-4645-a0e7-433ed0143b33" />

I don't understand what you mean by being consistent? I took the code from the very first CMS in the list and adopted it for ProcessWire.

tyler36 replied:
> I don't understand what you mean by being consistent? 

Almost all the other framework use `composer`; although there are a couple of exceptions.
Backdrop is a "special" case as it is a fork of Drupal _before_ they moved to a composer-based installation process; thus designed NOT to use composer.

BernhardBaumrock replied:
Thx for the clarification @tyler36

I have tried your suggested version and it works. The warning mentioned above is not nice to see though. Is that something PW has to change or is that on ddev side?

I don't like that it creates a /vendor folder in the root directory that I don't need. I've been using PW for over 10 years now and never ever installed it via composer. I don't think many do. To be honest I'm not sure about your suggestion...

stasadev replied:
> The warning mentioned above is not nice to see though. Is that something PW has to change or is that on ddev side?

It should be a `project` and not a `library`?

https://github.com/processwire/processwire/blob/3cc76cc886a49313b4bfb9a1a904bd88d11b7cb7/composer.json#L3

```diff
-"type": "library",
+"type": "project",
```

See

- https://github.com/composer/composer/pull/11885

BernhardBaumrock commented:
That are some great points @rpkoller , thank you!

I've been using PW for over 10 years, so I'm blind for many things :)

> Talking of consistency with other quickstarts, i am not sure if i would provide instructions how to install a dev version at all. all other quickstarts provide either instructions to the last major version and or a git repo, and if no composer based install is available a zip archive based install approach. plus main branch and dev branch might be confusing as well. the general instructions  To get started with [ProcessWire](https://processwire.com/), clone the project repository and navigate to the project directory. in combination with the aforementioned tab labels provide sort of a mixed messaging cuz both available options are about downloading a zip archive and not to git cloning a repo.

Not sure I understand 100% but I get the point that instructions to use "git clone" are misleading when PW is later installed via composer, thx. I've added the suggested wording.

> when i install processwire i get the following warning in the context of ddev

I get that too and I simply ignore it. Good point to add it. I have updated the readme with all suggestions I got so far plus I added a link to the PW forum thread about DDEV.

rfay commented:
This will need paired automated tests, like we have now done with Backdrop and intend to do with everything. Are you interested in adding those here @BernhardBaumrock ? I know it will be more than you intended to do. It's not terribly hard, see the example in https://github.com/ddev/ddev/tree/master/docs/tests

rfay commented:
Converted to draft, needs tests to be pulled.

BernhardBaumrock commented:
Sorry @rfay I don't understand. Why did my last commit fail? It had only 2 typos fixed. Why do we need tests? Do they make sure that the docs are correct and you get notified if something breaks?

rfay commented:
`ProcessWire` needs to be added to the spellcheck list in .spellcheckwordlist.txt

https://github.com/ddev/ddev/actions/runs/12634221365/job/35203193464?pr=6879

rfay commented:
> Sorry @rfay I don't understand. Why did my last commit fail? It had only 2 typos fixed. Why do we need tests? Do they make sure that the docs are correct and you get notified if something breaks?

We find that upstream changes often end up making the quickstarts invalid, and we don't have a way to know about that, so then people come along needing support.

BernhardBaumrock commented:
So what you are asking for is to write an automated test that will make sure that the instructions from the quickstart work.

And if the upstream (processwire) changed anything for whatever reason that would break the instructions on the quickstart you'd get notified and avoid support requests and can do something about it instantly. Correct?

BernhardBaumrock commented:
> ProcessWire needs to be added to the spellcheck list in .spellcheckwordlist.txt

Are you asking me to do that or are you just dropping notes? If that was towards me I'd appreciate some more information on how to do that. Thx!

rfay commented:
Hi @BernhardBaumrock - 

* For this to pass tests, it will have to pass the spellcheck. 
* You're the owner of this PR, so it would be lovely if you'd push an additional commit to do that
* You're invited to do the automated tests for this PR/Processwire, but if you don't somebody else will before it gets pulled.
* @stasadev opened https://github.com/ddev/ddev/issues/6880 so we can do more extensive instruction on how to do this and in general how to add a new quickstart.

> So what you are asking for is to write an automated test that will make sure that the instructions from the quickstart work.

> And if the upstream (processwire) changed anything for whatever reason that would break the instructions on the quickstart you'd get notified and avoid support requests and can do something about it instantly. Correct?

Yes, this is correct. But if you don't want to take on this challenge, we'll get it done. We only recently started adding automated tests for quickstarts. This is the first one that has triggered this requirement. But we're planning to add tests for all of them.

BernhardBaumrock commented:
thx @rfay 

> For this to pass tests, it will have to pass the spellcheck.

I understand that, but I have no idea how to make it pass the spellcheck. It might be obvious to you but it is not for me ;)

I'll wait for the FAQ on how to do it properly. Thx.

Anybody else feel free to beat me ;)

rfay commented:
If you're familiar with git, check out your branch (https://github.com/BernhardBaumrock/ddev/tree/patch-2). Add a commit to it that adds "ProcessWire" to the file [.spellcheckwordlist.txt](https://github.com/ddev/ddev/blob/master/.spellcheckwordlist.txt). If you want support or coaching on how to do that in git, we're here to help, join us on [discord](https://ddev.com/s/discord)

rfay commented:
BTW, I just want to say how much I've appreciated your work in 
* https://github.com/processwire/processwire-issues/issues/1961

although it does seem like they don't understand...

BernhardBaumrock commented:
@rfay thank you. Yeah, it's unfortunate, but I can live with that. DDEV is still an outstanding project and I created a snippet for me that adds the relevant lines to the config as soon as I type :dirs 

Regarding my PR: I removed Simply from my text, but I also found "simply" used in the Contao text. Not sure if that is missing in your checks or if the checks only scan newly added lines of text?

rfay commented:
Rebased.

rfay commented:
When going through manually and adding the test I fiddled with a few things.
* `ddev start` is good for people to know about so I put it in there.
* utf8mb4 is needed for a successful website in 2024 (and it was in 2014) otherwise all emojis cause crazy data loss.
* DDEV uses InnoDB, MyISAM hasn't been used anywhere I know of in the last 10-15 years. 
* `processwire/processwire:^3` needed to be in quotes because the caret is a special character in PowerShell

I'm a little worried about the composer technique being listed first. If folks using ProcessWire actually never use composer then that's not exactly what we want. Do they usually use the tarball download technique that you initially suggested?

BernhardBaumrock commented:
Hey @rfay thank you very much for your work on this!

I agree with all what you say.

> utf8mb4 is needed for a successful website in 2024 (and it was in 2014) otherwise all emojis cause crazy data loss.

I agree, but I'm not sure what you want to say? Do you see an issue here? Anything that we have to fix / work on?

> DDEV uses InnoDB, MyISAM hasn't been used anywhere I know of in the last 10-15 years.

Same as above. Does that mean PW + DDEV will only work with "InnoDB", right? I've never tried a different setting for years, as PW supports and recommends InnoDB since 2018 (https://processwire.com/blog/posts/using-innodb-with-processwire/)

> I'm a little worried about the composer technique being listed first. If folks using ProcessWire actually never use composer then that's not exactly what we want. Do they usually use the tarball download technique that you initially suggested?

And I'd absolutely prefer it showing the non-composer way of installing PW. From my experience and what I see in the forum only very few advanced users use composer to install it. I think the majority uses ZIP, yes.

Please let me know if there is anything else from my/our side!

rpkoller commented on docs/tests/processwire.bats:
```diff
@@ -0,0 +1,37 @@
+  # validate running project
+  run curl -sfI https://${PROJNAME}.ddev.site
+  assert_success
+  assert_output --partial "server: Apache"
```
```suggestion
  assert_output --partial "server: Apache"
  assert_output --partial "HTTP/2 200"
```

rpkoller replied:
i would suggest to assert the http response status as well, same as on the other bats test

rpkoller commented on docs/content/users/quickstart.md:
````diff
@@ -629,6 +629,53 @@ The Laravel project type can be used for [StarterKits](https://laravel.com/docs/
+- **After installation,** configure `upload_dirs` to specify where user-generated files are managed by Processwire:
+
+    ```
+    ddev config --upload-dirs=sites/assets/files` && ddev restart
````
```suggestion
    ddev config --upload-dirs=sites/assets/files && ddev restart
```

BernhardBaumrock replied:
Isn't the upload_dirs only necessary when mutagen is enabled, which is not always the case? Not sure if it is a problem if the setting is there but mutagen is disabled?

rpkoller replied:
the problem is the single quote that got removed in my recommendation. with the single quote in the command isnt working.

BernhardBaumrock replied:
@rfay I understand what this fix is about, but seeing it I realised that you changed my text as well and it now does not mention that this setting is only needed when using mutagen. I had that description in my text. But I don't know if that is a problem or not, I just wanted to mention it for you guys so you can decide and maybe change it or leave it.

rfay replied:
`upload_dirs` is not just for mutagen users. It is how `ddev import-files` works, how it knows where to put the files.

BernhardBaumrock replied:
Thx for the clarification. Seems I still don't get what it does, but I don't really care. DDEV needs to work for me and it does that.

Regarding upload_dirs this is the warning that I get when using DDEV for ProcessWire development without the upload_dirs set:

```
You have Mutagen enabled and your 'php' project type doesn't have `upload_dirs` set.
For faster startup and less disk usage, set upload_dirs to where your user-generated files are stored.
If this is intended you can disable this warning with `ddev config --disable-upload-dirs-warning`.
```

Which is why I thought it is related to mutagen...

BernhardBaumrock replied:
I have never used / needed ddev import-files ...

rfay replied:
There are two good ways to get rid of the warning. 

* As it says in the message, you can disregard with `ddev config --disable-upload-dirs-warning`

* But it's far better (assuming ProcessWire does put user-generated files in a consistent place) to name that place in the upload_dirs configuration.

BernhardBaumrock replied:
Yes, whenever a user uploads a file to the CMS it will be put into /site/assets/files/123/my-upload.jpg

rfay replied:
> I have never used / needed ddev import-files

Please remember that this quickstart is not just for you, it's for all your friends in ProcessWire.

BernhardBaumrock replied:
@rfay this comment is absolutely unnecessary. Why are you implying that? I'm just reporting my experience and I try to help adding a new system to the docs.

It's your job to collect the information that you get and then make the best out of it. Obviously you are the experts about DDEV and I am not. If we talked about ProcessWire it would be the other way round. I can just share my experience and what I did do and also what I did not do. So that you, with your knowledge, can decide what needs to be written to the docs and what can be ignored. Sometimes even unexperienced users shed some fresh light on things where experienced users got blind over time. If that was not the case... Well... Fine. There is no guarantee that what I report and contribute is helpful. If not, just ignore it.

But accusing me of not thinking about others (which I interpret as a very rude statement) is really not professional.

In fact it is the opposite. I wrote a comment about that section about "upload_dirs" because I was unsure if that is actually a generic description that is valid for everybody, not just for me (or anybody else using mutagen).

Also... I've been using DDEV for maybe two years now. And I'm very happy with it and I'm very thankful for what you are doing here. But why do you think am I trying to add ProcessWire to the docs? For myself? Seriously?

But maybe I'm reading it wrong. Please remember that not everyone's mother tongue is english, so some details might got lost (or added) in translation.

rfay replied:
You seem to be arguing over and over again in multiple places that `upload_dirs` is not relevant. I've answered and explained many times. It's relevant. And if ProcessWire is consistent, we should set it. I'm unable to understand why you keep arguing about it. Your comment implied to me that "Because I haven't needed it it's not relevant".

BernhardBaumrock replied:
> You seem to be arguing over and over again in multiple places that upload_dirs is not relevant.
Please show me a single place where I stated that.

Other than that please let me know if you need any input from my side to get this merged. But to be honest I'd prefer to close this PR.

Just don't want to do anything that might cause more work on your side.

rfay replied:
I'm sorry you're frustrated. I am too. But it doesn't make sense to throw away both your work and mine. We'll get to it. Why don't you ignore it for now and come back and do an update PR after it goes in?

rfay reviewed (approved):
Thanks for the contribution and all the great work on this!

stasadev commented on docs/content/users/quickstart.md:
```diff
@@ -640,6 +640,66 @@ The Laravel project type can be used for [StarterKits](https://laravel.com/docs/
+
+To get started with [ProcessWire](https://processwire.com/), create a new directory and use the Zip file download, composer, or Git checkout to build. These instructions are adapted from [ProcessWire Install Documentation](https://processwire.com/docs/start/install/new/#installing-processwire).
+
+=== "Zip File"
```
```suggestion
=== "ZIP File"
```

stasadev commented on docs/content/users/quickstart.md:
```diff
@@ -640,6 +640,66 @@ The Laravel project type can be used for [StarterKits](https://laravel.com/docs/
 
+## ProcessWire
+
+To get started with [ProcessWire](https://processwire.com/), create a new directory and use the Zip file download, composer, or Git checkout to build. These instructions are adapted from [ProcessWire Install Documentation](https://processwire.com/docs/start/install/new/#installing-processwire).
```
```suggestion
To get started with [ProcessWire](https://processwire.com/), create a new directory and use the ZIP file download, composer, or Git checkout to build. These instructions are adapted from [ProcessWire Install Documentation](https://processwire.com/docs/start/install/new/#installing-processwire).
```

stasadev reviewed (approved):
Looks good to me, thanks!

BernhardBaumrock commented:
Thx
